Changing SnackBar's default vertical padding

rickimaru18 opened the pull request:
Make "_singleLineVerticalPadding" a changeable property of SnackBar.

Fixes: #77333

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

rickimaru18 committed:
Changing SnackBar's default vertical padding

Make "_singleLineVerticalPadding" a changeable property of SnackBar.

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -15,8 +15,6 @@ import 'text_button.dart';
 import 'text_button_theme.dart';
 import 'theme.dart';
 
-const double _singleLineVerticalPadding = 14.0;
```
We would probably want to keep this constant so we know it is the default rather than just a number in the constructor.

rickimaru18 replied:
@Piinks  Understood :bow:

rickimaru18 replied:
Committed: 739f429

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -367,6 +366,16 @@ class SnackBar extends StatefulWidget {
+  /// This will not be overriden when setting [padding]'s top and/or bottom values.
+  /// Both paddings will just be added.
+  /// 
+  /// To ignore this padding, please set to 0.
```
This comment does not seem necessary.

rickimaru18 replied:
I'll just wait for your response in this comment: https://github.com/flutter/flutter/pull/78133#discussion_r596321803
:bow:

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -367,6 +366,16 @@ class SnackBar extends StatefulWidget {
+  /// 
+  /// Defaults to 14.0.
+  /// 
+  /// This will not be overriden when setting [padding]'s top and/or bottom values.
+  /// Both paddings will just be added.
```
Can you elaborate on this a bit more? I am not sure I understand the expected behavior from this.

rickimaru18 replied:
Because `SnackBar` also have a `padding` property. So, these two paddings will just add up.

Sample:
```
padding: const EdgeInsets.symmetric(vertical: 10),
singleLineVerticalPadding: 20,
```
Total vertical padding will be 30.

Piinks replied:
Hm, this makes me wonder if we should add this. It appears to be opportune for confusion. Could we not achieve the desired effect by using the already exposed `padding`?

rickimaru18 replied:
@Piinks 
Yes. I was also thinking that it would be better to just use the `padding` property. I'll commit a version for this. :bow:

rickimaru18 replied:
@Piinks Committed: 9b459a5

Piinks reviewed (commented):
Hi @rickimaru18 welcome, and thanks for the contribution! This will need a test to verify that the expected behavior works.

rickimaru18 committed:
Keep "_singleLineVerticalPadding" constant so we know it is the default rather than just a number in the constructor.

rickimaru18 committed:
Use "padding" property to override default vertical padding of SnackBar

- Removed "singleLineVerticalPadding" property because this can be done by just setting the "padding" property.
- In "Snackbar padding can be customized" test, removed adding of default vertical padding.

Piinks reviewed (commented):
I am going to run this through google testing to see if this kind of change breaks anything - the simplest fixes have a knack for wreaking havoc 😉 
Can you check the failing tests here? It looks like your branch may just need to be updated.

rickimaru18 commented:
@Piinks It may destroy the UI for the SnackBars that depends on the 

> I am going to run this through google testing to see if this kind of change breaks anything - the simplest fixes have a knack for wreaking havoc 😉

It may destroy the UI for SnackBars that depends on the `_singleLineVerticalPadding`.

> Can you check the failing tests here? It looks like your branch may just need to be updated.

Understood.

rickimaru18 committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -504,7 +504,7 @@ class _SnackBarState extends State<SnackBar> {
           Expanded(
             child: Container(
-              padding: const EdgeInsets.symmetric(vertical: _singleLineVerticalPadding),
+              padding: widget.padding == null ? const EdgeInsets.symmetric(vertical: _singleLineVerticalPadding) : null,
```
This behavior is not documented anywhere. The documentation of `SnackBar.padding` applies to the SnackBar's contents, why would it affect the outside padding of the SnackBar?

rickimaru18 replied:
> why would it affect the outside padding of the SnackBar?

This is because we removed the `singleLineVerticalPadding` property, https://github.com/flutter/flutter/pull/78133#discussion_r604277207, to not confuse the users on which property can override the default vertical padding.

I forgot to update the documentation of `SnackBar.padding`. If my latest commit is okay, I'll proceed updating the documentation. 🙇

rickimaru18 replied:
> So SnackBar.padding wraps the entire contents of the SnackBar (content & action)

Yes.

> and this would only wrap the SnackBar.content (no action)

Yes. But only a vertical padding, `_singleLineVerticalPadding`.


> So if the user sets SnackBar.padding to go around the content and action, there will be 0 padding in between the content and the action?

The latest commit will only set vertical padding of the content to `null` but there's still a padding in between the content and action (horizontal padding).

Example:
```
SnackBar(
  content: const Text('The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.'),
  action: SnackBarAction(label: 'Sample Action', onPressed: () {}),
  padding: const EdgeInsets.all(0),
),
```
Will display...
<img width="358" alt="image" src="https://user-images.githubusercontent.com/13028267/118425077-fc8b5100-b6fa-11eb-81df-ee4117f1878c.png">

Piinks replied:
I think that makes sense. Would you like to update the documentation?

rickimaru18 replied:
@Piinks Is this update, 9c4ee1b, okay?

rickimaru18 committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

rickimaru18 committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

Piinks commented:
(triage) @rickimaru18 would you like to continue with this change?

rickimaru18 commented:
@Piinks 

Sorry for the late reply. I'll try to update the doc comments for `padding`. :bow:

rickkrystiannelim committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

rickkrystiannelim committed:
Update doc comments for "padding" property.

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -304,10 +304,11 @@ class SnackBar extends StatefulWidget {
-  /// the presence of an [action]. The start padding is 24 if [behavior] is
-  /// [SnackBarBehavior.fixed] and 16 if it is [SnackBarBehavior.floating]. If
-  /// there is no [action], the same padding is added to the end.
+  /// If this property is null, the vertical padding's default value will be 14,
```
This may be easier to read with if it broke down each padding aspect, like:

If this property is null, the default padding values are
- vertical..
- horizontal..
- etc.

I think it would also be beneficial to mention if it is padding around the content, the action or both, since that was difficult to understand while we were going back and forth on this. :)

rickimaru18 replied:
@Piinks 

I updated the documentation. 7f56a9a

Just realized that the `padding` implementation is somewhat complicated to explain because both `content` and `action` uses the same `padding` property.

rickimaru18 committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

rickimaru18 committed:
Update SnackBar "padding" property's documentation.

rickimaru18 committed:
Update "right padding" assignment documentation.

rickimaru18 committed:
Merge branch 'master' into snackbar_singlelintverticalpadding

rickimaru18 committed:
Removed extra trailing spaces.

Piinks reviewed (approved):
LGTM
